pkg/binding/controller.go

```diff
@@ -115,10 +121,19 @@ func NewController(parentLogger logr.Logger, wdsRestConfig *rest.Config, imbsRes
 		return nil, err
 	}
 
-	nGVRs, err := countGVRsViaAggregatedDiscovery(kubernetesClient)
+	// do the discovery, save the results, then discard the 'depleted' client
```

Comment:
k8s.io/client-go maintains an HTTP/2 connection per _what_?

Reply:
Let me explain why a `disposableClient` is used here.

Before this PR, we pass binding controller's `kubernetesClient` to `func countGVRsViaAggregatedDiscovery` to do the discovery. After the discovery, the tokens could be 'depleted'. This depletion can be demonstrated by the attached cm.log file in the openning comment. If we look at lines 41 and lines 69 of the log, we can see that the `kubernetesClient` does two consecutive full (before this PR) discovery when applying KS CRDs. The first full discovery succeeded without showing throttling in the log (line 42). The second full discovery, however, triggers a group of throttling (starting with line 43), and leads to the context timeout, thus failure of start. The two full discoveries run the same section of code, but only the second discovery triggers throttling, that's because the first full discovery, perhaps plus part of the second full discovery, depleted `kubernetesClient`'s token bucket.

In this PR, `disposableClient` is assigned to the one-time full discovery. After the heavy-lifting job, `disposableClient` could be (based on # GVRs in WDS) depleted, but it is not returned by `NewController`, thus pop from the stack and disposed. By doing this, we make sure that binding controller's `kubernetesClient` is fresh with a full token bucket. This fresh `kubernetesClient` is now ready to do the two consecutive discoveries when applying CRDs. Also, the two consecutive discoveries are not full discovery anymore, they are just very light discoveries in one single GV.

Reply:
I was reacting to the comment, questioning the implicit assumption about what the rate limiter is associated with; how do you know that the code in this PR causes two different rate limiters to be used? Certainly the HTTPS connection is shared (see https://github.com/kubernetes/client-go/blob/v0.28.2/transport/transport.go#L54).

I realize now that I mistakenly asked about the HTTPS connection when I _should_ have asked about the rate limiter. I have since looked, and found the answer: if you do not set the `RateLimiter` field of the `rest.Config` then you get a new rate limiter with each call to `kubernetes.NewForConfig` (regardless of what `*rest.Config` is passed --- N.B. that is a pointer type). You can see this at https://github.com/kubernetes/client-go/blob/v0.28.2/kubernetes/clientset.go#L473 and https://github.com/kubernetes/client-go/blob/v0.28.2/kubernetes/clientset.go#L481-L486 .

So, in short, I have answered my question about the comment. It is accurate.

Reply:
Thanks for the pointers! BTW, I am a bit confused about the word 'shallow' in https://github.com/kubernetes/client-go/blob/6030d798e44096d5ffb58e00802cadd81a74dcf0/kubernetes/clientset.go#L461. By creating a variable `configShallowCopy` after the dereference, don't we have a 'deep' copy of the content that `c` points to?